### PR TITLE
Feature/issue 191 gp squared exponential

### PR DESCRIPTION
**Pull request by rtrangucci**
#### Submisison Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

Add cov_exp_quad function to stan math

#### Intended Effect:

Reduce the user-written code for covariance matrices used in Gaussian Process models

#### How to Verify:

Run unit tests:

test/unit/math/fwd/mat/fun/squared_distance_test.cpp
test/unit/math/prim/scal/fun/squared_distance_test.cpp
test/unit/math/rev/scal/fun/squared_distance_test.cpp

test/unit/math/rev/mat/fun/cov_exp_quad_test.cpp
test/unit/math/prim/mat/fun/cov_exp_quad_test.cpp
test/unit/math/mix/mat/fun/cov_exp_quad_test.cpp

#### Side Effects:

None

#### Documentation:

Need to create doc in stan reference, and stan language

#### Reviewer Suggestions: 

@syclik 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Rob Trangucci

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bgoodri** [comment]
Shouldn't we also have a version that inputs a symmetric distance matrix? If the distances are fixed, then we don't need to be recalculating them at every evaluation.

File: stan/math/prim/mat/fun/cov_exp_quad.hpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+     * @return squared distance
+     * @throw std::domain_error if sigma <= 0, l <= 0, or
+     *   x is nan or infinite
+     */
+    template<typename T_x, typename T_sigma, typename T_l>
+    inline typename
+    Eigen::Matrix<typename stan::return_type<T_x, T_sigma, T_l>::type,
+                  Eigen::Dynamic, Eigen::Dynamic>
+    cov_exp_quad(const std::vector<T_x>& x,
+               T_sigma& sigma,
```
Comment (syclik) [inline stan/math/prim/mat/fun/cov_exp_quad.hpp]: fix indentation

File: stan/math/prim/mat/fun/cov_exp_quad.hpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+     * @return squared distance
+     * @throw std::domain_error if sigma <= 0, l <= 0, or
+     *   x is nan or infinite
+     */
+    template<typename T_x1, typename T_x2, typename T_sigma, typename T_l>
+    inline typename
+    Eigen::Matrix<typename stan::return_type<T_x1, T_x2, T_sigma, T_l>::type,
+                  Eigen::Dynamic, Eigen::Dynamic>
+    cov_exp_quad(const std::vector<T_x1>& x1,
+               const std::vector<T_x2>& x2,
```
Comment (syclik) [inline stan/math/prim/mat/fun/cov_exp_quad.hpp]: fix indentation

**Comment by syclik** [comment]
Looks good to me. Just a couple of cosmetic things to fix before merging. If it doesn't get touched in the next couple days, I'll handle it.

@bgoodri, can you create a new issue for that input? No need to tack it on this pull request.

**Comment by rtrangucci** [comment]
@syclik I'll take care of your comments this afternoon.

**Comment by syclik** [comment]
Thanks!

**Comment by syclik** [comment]
Nah. That reminder was good enough!

**Pull request merged**